lxd/storage/drivers/drivers_lvm_volumes.go

```diff
@@ -0,0 +1,732 @@
+			return err
+
+		}
+	}
+
+	return nil
+}
+
+// CreateVolumeSnapshot creates a snapshot of a volume.
+func (d *lvm) CreateVolumeSnapshot(snapVol Volume, op *operations.Operation) error {
+	parentName, _, _ := shared.InstanceGetParentAndSnapshotName(snapVol.name)
+	parentVol := NewVolume(d, d.name, snapVol.volType, snapVol.contentType, parentName, snapVol.config, snapVol.poolConfig)
+	snapPath := snapVol.MountPath()
+
+	// Create the parent directory.
+	err := createParentSnapshotDirIfMissing(d.name, snapVol.volType, parentName)
+	if err != nil {
+		return err
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Create snapshot directory.
+	err = snapVol.EnsureMountPath()
+	if err != nil {
+		return err
+	}
+	revert.Add(func() { os.RemoveAll(snapPath) })
+
+	_, err = d.createLogicalVolumeSnapshot(d.config["lvm.vg_name"], parentVol, snapVol, true, d.usesThinpool())
+	if err != nil {
+		return fmt.Errorf("Error creating LVM logical volume snapshot: %v", err)
+	}
+
+	revert.Success()
+	return nil
+}
+
+// DeleteVolumeSnapshot removes a snapshot from the storage device. The volName and snapshotName
+// must be bare names and should not be in the format "volume/snapshot".
+func (d *lvm) DeleteVolumeSnapshot(snapVol Volume, op *operations.Operation) error {
+	// Remove the snapshot from the storage device.
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if lvExists {
+		_, err = d.UnmountVolume(snapVol, op)
+		if err != nil {
+			return fmt.Errorf("Error unmounting LVM logical volume: %v", err)
+		}
+
+		err = d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name))
+		if err != nil {
+			return fmt.Errorf("Error removing LVM logical volume: %v", err)
+		}
+	}
+
+	// Remove the snapshot mount path from the storage device.
+	snapPath := snapVol.MountPath()
+	err = os.RemoveAll(snapPath)
+	if err != nil {
+		return errors.Wrapf(err, "Error removing LVM snapshot mount path '%s'", snapPath)
+	}
+
+	// Remove the parent snapshot directory if this is the last snapshot being removed.
+	parentName, _, _ := shared.InstanceGetParentAndSnapshotName(snapVol.name)
+	err = deleteParentSnapshotDirIfEmpty(d.name, snapVol.volType, parentName)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RenameVolumeSnapshot renames a volume snapshot.
+func (d *lvm) RenameVolumeSnapshot(snapVol Volume, newSnapshotName string, op *operations.Operation) error {
+	volDevPath := d.lvmDevPath(d.config["lvm.vg_name"], snapVol.volType, snapVol.contentType, snapVol.name)
+	lvExists, err := d.logicalVolumeExists(volDevPath)
+	if err != nil {
+		return err
+	}
+
+	if !lvExists {
+		return fmt.Errorf("LVM logical volume doesn't exist")
+	}
+
+	_, err = d.UnmountVolume(snapVol, nil)
+	if err != nil {
+		return fmt.Errorf("Error unmounting LVM logical volume: %v", err)
+	}
```

Comment:
Hmm, if the snapshot is currently mounted, it's probably best to just completely abort the rename.

Though it feels like mediating such actions would be better handled by the backend. Effectively preventing any rename/delete/update while mounted.

Reply:
@tomponline got this one to address still

Reply:
@stgraber ive removed the unmount call, for the wider change my plan is:

1. Add a IsVolumeMounted(vol Volume) function to all drivers (and the driver interface).
2. Implement that function for each driver.
3. Check whether the volume is mounted in backendLXD, and if so, fail in: rename/delete volume operations
4. For backendLXD updates, in order to support online volume growth we need to push the request down into the driver to decide whether the change is a growth or a shrink and whether it can be accommodated. For LVM this will check for mounted and fail for any other operation except grow.